### DISCUSSION TITLE
Create handler file for fuzz testing or just a modifier?

### DISCUSSION THREAD

**Discussion by jaimebarrancos** [Cyfrin/foundry-full-course-cu#363] [Q&A] [answered]
Why do we create a handler file for fuzz testing? Wouldn't it make more sense to just create a modifier ?

**Comment by ghost** [comment, marked as answer]
My perspective:

Invariant tests are meant to test our protocol as a whole, not only specific properties of it. The problem is that the invariant test goes literally crazy creating random sequences of calls to break the system and we can end up with a lot of unreliable results because of the unrealistic ordering of the sequence of calls e.g. a call to a  withdraw method without depositing tokens first.

To improved the test campaign we can configure it with a Handler contract that will act as a wrapper. This contract will provide extra logic to govern the calls made to our protocol to ensure successful and reliable results e.g. now we deposit first and then we withdraw.

So basically:

The test calls the Handler's functions, then the Handler will call the methods in our protocol and finally the invariants are verified.  

The challenge is having good assumptions (easier said than done) of our protocol to design a good handler.

I think you can gain a lot more control over the execution of your invariant test by scaling from a modifier to an actual contract.
